File: typescript/src/cacheToken.ts
import dotenv from 'dotenv';
import fs from 'node:fs';

dotenv.config();

const TOKEN_CACHE_FILE = '.token-cache.json';

const BASE_URL = 'https://api.hyperswitchai.com';

async function getToken() {
  try {
    // Check if we have a cached token
    if (fs.existsSync(TOKEN_CACHE_FILE)) {
      const cache = JSON.parse(fs.readFileSync(TOKEN_CACHE_FILE, 'utf8'));
      // Check if token is still valid (with 5 minute buffer)
      if (cache.expiresAt > Date.now() + 300000) {
        return cache.token;
      }
    }

    // Get new token
    const authResponse = await fetch(`${BASE_URL}/auth`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        username: process.env.USERNAME,
        password: process.env.PASSWORD,
      })
    });

    if (!authResponse.ok) {
      throw new Error('Authentication failed');
    }

    const { token } = await authResponse.json() as { token: string };

    // Cache token with 1 hour expiry
    fs.writeFileSync(TOKEN_CACHE_FILE, JSON.stringify({
      token,
      expiresAt: Date.now() + 3600000 // 1 hour. Note that this is just for our convenience in determining if the token is still valid. Changing this value does not affect the token's actual expiration from the server.
    }));

    return token;
  } catch (error) {
    console.error('Auth error:', error);
    throw error;
  }
}

async function main() {
  try {
    const token = await getToken();

    console.log('\n\x1b[32m%s\x1b[0m', '✅ Token successfully cached in .token-cache.json');
    
    console.log('Token:', token);

    console.log('\x1b[36m%s\x1b[0m', 'Now the token is cached in .token-cache.json, you can use it in your code and not have to get a new one all the time (until it expires).');

    console.log('\x1b[36m%s\x1b[0m', 'If you run this code again you should see the same token is being used.');


  } catch (error) {
    console.error('Error:', error);
  }
}

main();

File: python/src/cache_token.py
import os
import json
import time
from pathlib import Path
from dotenv import load_dotenv
import requests

# Load environment variables from .env file
load_dotenv()

TOKEN_CACHE_FILE = '.token-cache.json'
BASE_URL = 'https://api.hyperswitchai.com'

def get_token():
    try:
        # Check if we have a cached token
        cache_path = Path(TOKEN_CACHE_FILE)
        if cache_path.exists():
            with cache_path.open('r') as f:
                cache = json.load(f)
                # Check if token is still valid (with 5 minute buffer)
                if cache['expiresAt'] > int(time.time() * 1000) + 300000:
                    return cache['token']

        # Get new token
        auth_response = requests.post(
            f'{BASE_URL}/auth',
            headers={'Content-Type': 'application/json'},
            json={
                'username': os.getenv('USERNAME'),
                'password': os.getenv('PASSWORD')
            }
        )
        
        # Check if request was successful
        auth_response.raise_for_status()
        
        # Extract token from response
        token = auth_response.json()['token']

        # Cache token with 1 hour expiry
        cache_data = {
            'token': token,
            'expiresAt': int(time.time() * 1000) + 3600000  # 1 hour. Note that this is just for our convenience in determining if the token is still valid. Changing this value does not affect the token's actual expiration from the server.
        }
        
        with cache_path.open('w') as f:
            json.dump(cache_data, f)

        return token

    except requests.exceptions.RequestException as e:
        print('Auth error:', str(e))
        raise
    except Exception as e:
        print('Auth error:', str(e))
        raise

def main():
    try:
        token = get_token()

        print('\n\033[92m✅ Token successfully cached in .token-cache.json\033[0m')
        print('Token:', token)
        
        print('\033[96mNow the token is cached in .token-cache.json, you can use it in your code and not have to get a new one all the time (until it expires).\033[0m')
        print('\033[96mIf you run this code again you should see the same token is being used.\033[0m')

    except Exception as e:
        print('Error:', str(e))

if __name__ == '__main__':
    main()
